content/blog/2018/app-metrics-istio/index.md

```diff
@@ -0,0 +1,83 @@
+
+One thing you need to pay attention to is the port where you’re exposing your Prometheus metrics. In ASP.NET Core, the default port is **5000**. When running locally, application metrics are exposed on `localhost:5000/metrics`. However, when you containerize your app, it’s common to expose your application over a different port, like **8080**, and this becomes relevant later when we talk about configuration.
+
+Assuming that you containerized and deployed your application on an Istio-enabled Kubernetes cluster, let’s now take a look at what we need to do to get these application metrics scraped by Istio’s Prometheus.
+
+## Configuration
+In Istio 1.0.5, the default installation files for Kubernetes, `istio-demo.yaml` or `istio-demo-auth.yaml`, already have scraping configurations for Prometheus under a `ConfigMap`. You can just search for `prometheus.yml`. There are 2 scraping jobs that are relevant for application metrics: 
+
+{{< text yaml >}}
+- job_name: 'kubernetes-pods'
+   kubernetes_sd_configs:
+   - role: pod
+...
+- job_name: 'kubernetes-pods-istio-secure'
+   scheme: https
+
+{{< /text >}}
+
+These are the jobs that scrape metrics from regular pods and pods where mTLS is enabled. It looks like Istio’s Prometheus should automatically scrape application metrics. However, in my first try, it didn’t work. I wasn’t sure what was wrong but Prometheus has some default endpoints:
+
+* `/config`: to see the current configuration of Prometheus.
+* `/metrics`: to see the scraped metrics.
+* `/targets`: to see the targets that’s being scraped and their status. 
+
+All of these endpoints are useful for debugging Prometheus:
+
+{{< image width="60%" ratio="43%" link="./image1.png" caption="Prometheus Targets page" >}}
+
+Turns out, I needed to add some annotations in my pod YAML files in order to get Prometheus scrape the pod. I had to tell Prometheus to scrape the pod and on which port with these annotations:
+
+{{< text yaml >}}
+template:
+   metadata:
+      annotations:
+         prometheus.io/scrape: "true"
+         prometheus.io/port: "8080"
+{{< /text >}}
+
+After adding the annotations, I was able to see my application’s metrics in Prometheus: 
+
+{{< image width="60%" ratio="43%" link="./image2.png" caption="Prometheus page with application metrics" >}}
+
+However, it was only working for regular pods and I was not able to see metrics with mTLS enabled between pods.
+
+## Bug with Istio Prometheus
```

Comment:
i'm concerned about calling this "Istio Prometheus".  We don't offer an Istio-specific build/version of Prometheus. Rather, we provide a standard Prometheus install with Istio-specific configuration -- just as anyone running their own Prometheus instance is probably already doing.

I might call this `Issue with using Istio certs with Prometheus`.

So I would, `When the Istio Prometheus server starts` should become `When Prometheus starts` (and I'd suggest something like: `When Prometheus starts, it will attempt to mount the Istio-supplied certificates. However, they may not have been issued by Istio Citadel yet. Unfortunately, Prometheus does not retry to load the certificates, which leads to an issue scraping mTLS-protected endpoints.`)

Reply:
Fixed accordingly,